lib/rules/prefer-const.js

```diff
@@ -203,6 +205,128 @@ function groupByDestructuring(variables, ignoreReadBeforeAssign) {
+
+        for (let i = 0; i < numberOfElements; i++) {
+            if (!node.elements[i].elements && node.elements[i].type === "Identifier") {
+                identifiers.push(node.elements[i]);
+            } else {
+                const innerIdentifiers = getIdentifiersInArrayDestructureGroup(node.elements[i]);
+
+                innerIdentifiers.forEach(identifierNode => identifiers.push(identifierNode));
+            }
+        }
+    }
+    return identifiers;
+}
+
+/**
+ * Returns a count of nodes
+ * This will count nested nodes in nodes elements attribute.
+ *
+ * @param {ASTNode} node - node to count elements.
+ * @returns {integer} count nodes.
+ */
+function countElementsInArrayDestructureGroup(node) {
+    let count = 0;
+
+    if (node.elements) {
+        const numberOfElements = node.elements.length;
+
+        for (let i = 0; i < numberOfElements; i++) {
+            if (!node.elements[i].elements) {
+                count += 1;
+            } else {
+                count += countElementsInArrayDestructureGroup(node.elements[i]);
+            }
+        }
+    }
+    return count;
+}
+
+/**
+ * Returns a list of nodes
+ * that have have the value type Identifier.
+ * This will search for nested Identifiers.
+ *
+ * @param {ASTNode} node - node to search properties that have value Identifier.
+ * @returns {ASTNode[]} list Identifier nodes.
+ */
+function getIdentifiersInObjectDestructureGroup(node) {
+    const identifier = [];
+    const propertiesLength = node.properties.length;
+
+    for (let i = 0; i < propertiesLength; i++) {
+        if (node.properties[i].value.type === "Identifier") {
+            identifier.push(node.properties[i].value);
+        }
+    }
+    return identifier;
+}
+
+/**
+ * Returns a count of nodes
+ * This will count the node properties length
+ *
+ * @param {ASTNode} node - node to count properties.
+ * @returns {integer} count properties length.
+ */
+function countElementsInObjectDestructureGroup(node) {
+    return node.properties.length;
+}
+
+/**
+ * Checks to see if there are less identifier nodes in a destructure group
+ * than the number of terms in the group. It might mean that there
+ * is a term in the group that cannot be converted to const.
+ * We want to make sure all terms can be reported
+ * If not, we should remove the terms that would be reported in "any" case
+ *
+ * @param {Map<ASTNode, ASTNode[]>} identifierMap - Variables to group by destructuring.
+ * @param {boolean} checkingMixedDestructuring - boolean to check if any value in destructuring
+ *      should use const
+ * @param {integer|null} destructureCount count of destructure terms.
+ * @param {ASTNode[]|null} destructureIdentifier list of Identifier nodes to check 
+ *      in IdentifierMap
+ * @returns {Map<ASTNode, ASTNode[]>} Grouped identifier nodes.
+ */
+function verifyAllDestructuring(identifierMap, checkingMixedDestructuring, destructureCount, destructureIdentifier) {
```

Comment:
Would it be possible for this function to return a new `Map` instead? In general I don't think it's a good idea for functions to mutate their arguments.